### PR TITLE
chore(nexus/channel): reconfigure only device which is involved in DR…

### PR DESCRIPTION
**Pull request by dsharma-dc**
… event

### CODE REVIEW

**Comment by dsharma-dc** [comment]
bors try

**Comment by dsharma-dc** [comment]
bors try

**Comment by dsharma-dc** [comment]
Opening for review. The extended test run of "change volumes and replicas" which disrupts channels quite a bit has been running on this change for more than 12hours.

File: io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs [hunk 1]
```diff
@@ -126,7 +126,10 @@ impl<'n> Nexus<'n> {
         // rebuilt would then need to be rebuilt again.
         // Ensuring that the dst child receives all frontend Write IO keeps all
         // rebuilt ranges in sync with the other children.
-        self.reconfigure(DrEvent::ChildRebuild).await;
+        self.reconfigure(DrEvent::ChildRebuildBegin(
+            self.get_child_device_name(dst_child_uri.as_str())?,
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:131]: In case of an error does it get cleaned up afterwards?

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 2]
```diff
@@ -49,14 +49,16 @@ pub enum IoMode {
...
-#[derive(Debug)]
+#[derive(Debug, Eq, PartialEq)]
 #[allow(clippy::enum_variant_names)]
 /// Dynamic Reconfiguration Events occur when a child is added or removed
 pub enum DrEvent {
     /// Child unplug reconfiguration event.
-    ChildUnplug,
-    /// Child rebuild event.
-    ChildRebuild,
+    ChildUnplug(String),
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_channel.rs:57]: What is this String here, a device name or a uri?
Seems its the device name, any reason for using that instead of the uri?

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 3]
```diff
@@ -193,67 +196,69 @@ impl<'n> NexusChannel<'n> {
...
+                            }
+                            Err(e) => {
+                                child.set_faulted_state(FaultReason::CantOpen);
+                                error!("{self:?}: failed to get I/O handle for {child:?}: {e}");
+                            }
+                        }
+                        // Make sure the child is not in readers. Most likely
+                        // the child won't be in readers
+                        // already here.
+                        self.readers
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_channel.rs]: Should this check be done out of the if let Some() ?

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 4]
```diff
@@ -193,67 +196,69 @@ impl<'n> NexusChannel<'n> {
...
-                            "{self:?}: failed to get I/O handle \
-                                for {c:?}: {e}"
-                        );
+                }
+            }
+            DrEvent::ChildRebuildEnd(dev) => {
+                if let Some(child) =
+                    self.nexus().child_by_device_name(dev.as_str())
+                {
+                    if child.is_healthy() {
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_channel.rs]: Do we need to check if it's healthy here?

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 5]
```diff
@@ -193,67 +196,69 @@ impl<'n> NexusChannel<'n> {
...
+                        // TODO: assert that the device is_writer and not
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_channel.rs]: Should we check that child is in the self.writers, and if not don't add it to the readers?

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 6]
```diff
@@ -193,67 +196,69 @@ impl<'n> NexusChannel<'n> {
...
+                            }
+                            Err(e) => {
+                                child.set_faulted_state(FaultReason::CantOpen);
+                                error!("{self:?}: failed to get I/O handle for {child:?}: {e}");
+                            }
+                        }
                     }
-                });
+                }
+            }
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_channel.rs:270]: It seems that with this code now there's scope for duplicate entries in the lists, we should probably ensure that doesn't happen..

File: io-engine/src/bdev/nexus/nexus_child.rs [hunk 7]
```diff
@@ -1104,6 +1104,7 @@ impl<'c> NexusChild<'c> {
 
         let state = self.state();
         let is_destroying = self.is_destroying();
+        let child_dev = self.get_device_name().unwrap_or_default();
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_child.rs]: Seems we should leave it as Result/Option rather than default to empty string?

File: io-engine/src/bdev/nvmx/channel.rs [hunk 8]
```diff
@@ -489,7 +489,6 @@ impl NvmeControllerIoChannel {
 
 impl Drop for NvmeControllerIoChannel {
     fn drop(&mut self) {
-        trace!("I/O channel {:p} dropped", self.0.as_ptr());
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nvmx/channel.rs]: Why is this removed, seems useful to have? Or is it too verbose?

File: io-engine/src/bdev/nvmx/qpair.rs [hunk 9]
```diff
@@ -200,7 +200,6 @@ impl QPair {
         // multiple allocations of the same handle for the same thread, to make
         // sure we don't reconnect every time.
         if self.state() == QPairState::Connected {
-            trace!(?self, "I/O qpair already connected");
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nvmx/qpair.rs]: Why is this removed, seems useful to have?

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 2 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_channel.rs:57]: Yes it's the device name. I guess no particular reason. For one, the traversal code is more closer to `device` side rather than `child`, and secondly, device name looks a bit compact/short compared to child uri.

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 4 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: Yes because reconfigure can get called in case of rebuild job Completed/Failed/Stopped, all of these. This maps to the check that happens today as well here.

File: io-engine/src/bdev/nvmx/channel.rs (hunk 8 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nvmx/channel.rs]: This looks like a mistake after I had modified these traces during testing. Will put back.

File: io-engine/src/bdev/nvmx/qpair.rs (hunk 9 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nvmx/qpair.rs]: This looks like a mistake after I had modified these traces during testing. Will put back.

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 4 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: Then shouldn't we remove the device from readers and writers in case the child is not healthy?

Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: In case child is not healthy as a result of rebuild fail/stop, the unplug workflow will kick in anyway so we'll be back here again as `ChildUnplug`, which will do the needful.

Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: Isn't this different behaviour from what happens today, when rebuild fails the reconfigure removes the child from writers straight away?

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 3 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: Yes, that can be done.

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 5 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: we can check that but I think that should be an assertion (as TODO says), because it shouldn't be possible - that the rebuild finished and child is healthy but it's not a channel writer on this core?

File: io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs (hunk 1 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:131]: you mean the String containing device name? hmm, It should be bound to lifetime of DrEvent I guess, so when the TraverseCtx is destroyed, the contained DrEvent(with String) should be destroyed too. Beyond `nexus::reconfigure`, there is no error path present today.

Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:131]: btw, today it's possible that if a rebuild fails and unplug is triggered, it'll reconfigure that channels and then later the `on_rebuild_update() `will again do the same which is redundant I believe. Now with this change, as a result of rebuild failure, the unplug will reconfigure the channels and the `on_rebuild_update() `will bail out early because `get_child_device_name() `will fail  with `ChildNotOpen` error:

> failed to process rebuild update notification with error: Child nvmf://65.21.0.104:8420/nqn.2019-05.io.openebs:233532e2-4fe0-436f-9eaa-c4652bc53f58?uuid=233532e2-4fe0-436f-9eaa-c4652bc53f58 of nexus f8079f24-6fe5-4532-ac88-fa0cbc31cbb4 is not open

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 4 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: straight away in this case means via on_rebuild_update() callback from rebuild backend, but by that time unplug is much ahead and has already done the reconfigure. `on_rebuild_update`() calls `close_faulted` before reconfigure

File: io-engine/src/bdev/nexus/nexus_child.rs (hunk 7 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_child.rs]: hmm, we can. The handling would become more verbose I think. e.g. I'd need to skip calling `reconfigure()` if the child_dev is `None`? OR take an `Option` all the way from `reconfigure` and beyond.

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 6 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_channel.rs:270]: The check would probably be based on iterating comparing the device name between existing NvmeDeviceHandle object and newly acquired one? vector length based check may not be completely reliable.

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 5 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: It can happen because calling reconfigure is async, so things may change underneath you.

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 4 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_channel.rs]: unplug IIRC is based on a callback from SPDK, so it may happen before or after depending on spdk and timing.
We should remove it on reconfigure and keep same behaviour.

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 6 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_channel.rs:270]: I guess we could use the name from NvmeDeviceHandle?

File: io-engine/src/bdev/nexus/nexus_child.rs (hunk 7 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_child.rs]: keep as Option and if None then skip the block_on call below.
Or should we force a full reconfigure in this case to be safe? (same as we have in current code)

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 6 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_channel.rs:270]: Added `is_reader() `and `is_writer()` helper functions in NexusChannel and used here.

**Comment by dsharma-dc** [comment]
Another run of extended test framework's "Change Replicas and Volume" test has been running for more than 8 hours on this change.

**Comment by dsharma-dc** [comment]
bors try

File: io-engine/src/bdev/nexus/nexus_channel.rs (hunk 2 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_channel.rs:57]: Let's document this here then, just a String could be anything.

File: io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs (hunk 1 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:131]: No, I mean does the rebuild job get cleaned up by upper layers if we fail here?

File: io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs [hunk 10]
```diff
@@ -462,7 +465,10 @@ impl<'n> Nexus<'n> {
             }
         }
 
-        self.reconfigure(DrEvent::ChildRebuild).await;
+        self.reconfigure(DrEvent::ChildRebuildEnd(
+            self.get_child_device_name(child_uri)?,
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:471]: hmm what happens if we fail here, we don't reconfigure channels is this ok?

**Review by tiagolobocastro — APPROVED** [review]
lgtm though a lot more convoluted, probably should be reviewed by @mtzaurus and/or @dsavitskiy

File: io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs (hunk 10 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:471]: Yes it should be ok because that would mean device is unplugged underneath us and reconfiguration would've happened. If the device is not found for any other reason, I think we would already have a broken system in other ways.

Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:471]: Just wanted to clarify because now with this change, there's nothing to fixup the channels, if we somehow "leak" a handle in the lists, there's no more reconconfigure_all that would remove them.

File: io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs (hunk 1 above)
Comment (dsharma-dc) [reply io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:131]: Looks like we need to call `child.remove_rebuild_job() `if at all this fails here. Will that drop the `RebuildJobBackEnd` as well or shall we explicitly do `self.comms.send(None)` before calling `child.remove_rebuild_job()`

File: io-engine/src/bdev/nexus/nexus_bdev.rs [hunk 11]
```diff
@@ -602,29 +602,30 @@ impl<'n> Nexus<'n> {
...
-            sender,
-            |chan, _sender| -> ChannelTraverseStatus {
-                chan.reconnect_all();
+            (sender, event),
+            |chan, (_sender, dev_ctx)| -> ChannelTraverseStatus {
+                chan.reconfigure_channel(dev_ctx);
                 ChannelTraverseStatus::Ok
             },
-            |status, sender| {
+            |status, ctx| {
```
Comment (dsavitskiy) [inline io-engine/src/bdev/nexus/nexus_bdev.rs:619]: you can destruct the tuple here as well: (sender, _).

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 12]
```diff
@@ -193,67 +210,70 @@ impl<'n> NexusChannel<'n> {
...
-                .for_each(|c| match c.get_io_handle() {
-                    Ok(hdl) => {
-                        debug!(
-                            "{self:?}: connecting child device \
-                                in write-only mode: {c:?}"
-                        );
-                        writers.push(hdl);
+        match dev_ctx {
+            DrEvent::ChildRebuildBegin(dev) => {
+                if let Ok(child) = self.nexus().child_by_device(dev.as_str()) {
```
Comment (dsavitskiy) [inline io-engine/src/bdev/nexus/nexus_channel.rs:227]: better to use &

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 13]
```diff
@@ -193,67 +210,70 @@ impl<'n> NexusChannel<'n> {
...
+                    debug_assert!(!self.is_writer(dev.as_str()));
+                    if !self.readers.is_empty() && !self.is_writer(dev.as_str())
```
Comment (dsavitskiy) [inline io-engine/src/bdev/nexus/nexus_channel.rs:229]: "!self.is_writer(dev.as_str()" is always true here, why adding it as a condition?

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 14]
```diff
@@ -193,67 +210,70 @@ impl<'n> NexusChannel<'n> {
...
+                // the child won't be in readers already here.
+                self.readers
+                    .retain(|h| &h.get_device().device_name() != dev);
+            }
+            DrEvent::ChildRebuildEnd(dev) => {
+                if let Ok(child) = self.nexus().child_by_device(dev.as_str()) {
+                    debug_assert!(!self.is_reader(dev.as_str()));
+                    if child.is_healthy()
+                        && self.is_writer(dev.as_str())
+                        && !self.is_reader(dev.as_str())
```
Comment (dsavitskiy) [inline io-engine/src/bdev/nexus/nexus_channel.rs:253]: Again, "!self.is_reader(dev.as_str()" is always true here.

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 15]
```diff
@@ -141,6 +144,20 @@ impl<'n> NexusChannel<'n> {
         self.readers.len()
     }
 
+    /// Check if the child with given device name is a writer in this channel.
+    pub fn is_writer(&self, dev_name: &str) -> bool {
```
Comment (dsavitskiy) [inline io-engine/src/bdev/nexus/nexus_channel.rs:148]: Is this needed to be "pub"? I cannot find it is used outside this module.

File: io-engine/src/bdev/nexus/nexus_channel.rs [hunk 16]
```diff
@@ -193,67 +210,70 @@ impl<'n> NexusChannel<'n> {
...
                     }
-                    Err(e) => {
-                        c.set_faulted_state(FaultReason::CantOpen);
-                        error!(
-                            "{self:?}: failed to get I/O handle \
-                                for {c:?}: {e}"
-                        );
+                }
+                // Also make sure the child is not in readers. Most likely
+                // the child won't be in readers already here.
```
Comment (dsavitskiy) [inline io-engine/src/bdev/nexus/nexus_channel.rs:244]: If the child is most likely not to be find in reader, it should either be a warning or assert, if it is.

**Review by dsavitskiy — CHANGES_REQUESTED** [review]

**Comment by dsharma-dc** [comment]
Changes merged via #1506 to handle the situation differently.
